extension/dnsmasq/dnsmasq.js

```diff
@@ -1323,4 +1329,60 @@ module.exports = class DNSMASQ {
       log.error("Failed to clean up leftover config", err);
     }
   }
+
+  //save data in redis
+  //{mac:{ipv4Addr:ipv4Addr,name:name}}
+  //host: { ipv4Addr: '192.168.218.160',mac: 'F8:A2:D6:F1:16:53',name: 'LAPTOP-Lenovo' }
+  async setupLocalDeviceDomain(restart, hosts, isInit) {
+    const json = await rclient.getAsync(LOCAL_DEVICE_DOMAIN_KEY);
+    try {
+      let needUpdate = false;
+      const deviceDomainMap = JSON.parse(json) || {};
+      for (const host of hosts) {
+        //if user update device name and ip, should get them from redis not host
+        let hostName = await rclient.hgetAsync(hostTool.getMacKey(host.mac), "name");
+        let ipv4Addr = await rclient.hgetAsync(hostTool.getMacKey(host.mac), "ipv4Addr");
+        hostName = hostName ? hostName : hostTool.getHostname(host);
+        ipv4Addr = ipv4Addr ? ipv4Addr : host.ipv4Addr;
+        if (!deviceDomainMap[host.mac]) {
+          deviceDomainMap[host.mac] = {
+            ipv4Addr: ipv4Addr,
+            name: hostName,
+            ts: new Date() / 1000
+          }
+          needUpdate = true;
+        } else {
+          const deviceDomain = deviceDomainMap[host.mac];
+          if ((deviceDomain.name != hostName || deviceDomain.ipv4Addr != ipv4Addr)) {
+            needUpdate = true;
+            deviceDomain.ipv4Addr = ipv4Addr;
+            deviceDomain.name = hostName;
+            deviceDomain.ts = new Date() / 1000
+          }
+        }
+      }
+      if (needUpdate || isInit) {
+        await rclient.setAsync(LOCAL_DEVICE_DOMAIN_KEY, JSON.stringify(deviceDomainMap));
+        let localDeviceDomain = "", domainMap = {};
+        for (const key in deviceDomainMap) {
+          const deviceDomain = deviceDomainMap[key]
+          if (!domainMap[deviceDomain.name]) {
+            domainMap[deviceDomain.name] = deviceDomain
+          } else if (domainMap[deviceDomain.name] && domainMap[deviceDomain.name].ts < deviceDomain.ts) {
+            domainMap[deviceDomain.name] = deviceDomain
+          }
+        }
+        for (const key in domainMap) {
+          const domain = getCanonicalizedHostname(key.replace(/\s+/g, ".")) + '.lan';
+          localDeviceDomain += `address=/${domain}/${domainMap[key].ipv4Addr}\n`
+        }
+        await fs.writeFileAsync(LOCAL_DEVICE_DOMAIN, localDeviceDomain);
+      }
+      if (needUpdate && restart) {
+        this.restart()
```

Comment:
Do not use this.restart(). Use this.restartDnsmasq() instead. It can introduce a cooldown time between each restart

Reply:
I use this.restartDnsmasq() before, but it looks like cannot work when I update device name.

Reply:
Figure out the issue. I used dnsmaq in Netbot.js before, it is in FireApi progress. So restartDnsmasq can not work. Refactor code in Netbot.js, trigger device update by the event.